feat: metric selection for faster calculation

tomsail opened the pull request:
Added new features: 
 * keyword `metrics` which can be `str` or `list`. Default `None` selects all metrics. 
 * added more thorough checks in tests
 * added duration print in pytest which gives the time spent for each function: 
 ```
 
=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/tomsail/work/python/seareport_org/seastats/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/tomsail/work/python/seareport_org/seastats
configfile: pyproject.toml
plugins: anyio-4.4.0
collected 5 items                                                                                                                                                                                                  

tests/compute_stats_test.py::test_get_all_stats_and_extremes PASSED                                                                                                                                          [ 20%]
tests/compute_stats_test.py::test_get_all_stats PASSED                                                                                                                                                       [ 40%]
tests/compute_stats_test.py::test_get_rmse PASSED                                                                                                                                                            [ 60%]
tests/compute_stats_test.py::test_get_slope_pp PASSED                                                                                                                                                        [ 80%]
tests/compute_stats_test.py::test_get_storm_metrics PASSED                                                                                                                                                   [100%]

================================================================================================ slowest durations =================================================================================================
0.26s call     tests/compute_stats_test.py::test_get_all_stats_and_extremes
0.24s call     tests/compute_stats_test.py::test_get_all_stats
0.07s call     tests/compute_stats_test.py::test_get_slope_pp
0.01s call     tests/compute_stats_test.py::test_get_storm_metrics
0.01s call     tests/compute_stats_test.py::test_get_rmse
0.00s setup    tests/compute_stats_test.py::test_get_all_stats_and_extremes
0.00s teardown tests/compute_stats_test.py::test_get_all_stats_and_extremes
0.00s teardown tests/compute_stats_test.py::test_get_all_stats
0.00s teardown tests/compute_stats_test.py::test_get_storm_metrics
0.00s teardown tests/compute_stats_test.py::test_get_rmse
0.00s setup    tests/compute_stats_test.py::test_get_all_stats
0.00s teardown tests/compute_stats_test.py::test_get_slope_pp
0.00s setup    tests/compute_stats_test.py::test_get_storm_metrics
0.00s setup    tests/compute_stats_test.py::test_get_rmse
0.00s setup    tests/compute_stats_test.py::test_get_slope_pp
```

NB: Calculating percentiles is the most intensive calculation, even more than POT selection.

tomsail committed:
feat: new API and simpler / faster calculations

pmav99 commented on seastats/stats.py:
```diff
@@ -1,5 +1,8 @@
 from typing import Dict
+from typing import List
```
on modern python versions (3.9+) builtin types are generics. This means that you don't need to import `typing.List` but you can just do `list[str]` instead. On older python versions you can still use the builtin functions as generics but you need to add `from __future__ import annotations` at the beginning of the file

https://mypy.readthedocs.io/en/latest/runtime_troubles.html#using-generic-builtins

pmav99 commented on seastats/stats.py:
```diff
@@ -178,6 +181,7 @@ def get_slope_intercept_pp(
 def get_stats(
     sim: pd.Series,  # The simulated time series data.
     obs: pd.Series,  # The observed time series data.
+    metrics: Optional[Union[str, List[str]]] = None,
```
So, this can be rewritten as 
```diff
-        metrics: Optional[Union[str, List[str]]] = None,
+        metrics: str | list[str] | None = None,
```
and you can remove the relevant imports.

That being said, and perhaps more importantly, as far as API design goes I would suggest the following instead:

``` python
from collections.abc import Sequence

_ALL_METRICS = ["bias", "rmse", "rms", ...]

def get_stats(
    sim: pd.Series,  # The simulated time series data.
    obs: pd.Series,  # The observed time series data.
    metrics: Sequence[str] = ["all"],
    )
    if metrics == ["all"]:
        metrics = _ALL_METRICS
    ...
    
```

This way the `metrics` argument is cleaner because it only accepts a single type and the implementation becomes simpler.

pmav99 replied:
PS. AI can be of great help when it comes to writing docs, but it tends to be overly verbose when it comes to comments. Please remove the ones that don't offer any value (e.g. the ones describing the arguments). If there is useful info there then it should be part of the docstring text.

pmav99 commented on tests/compute_stats_test.py:
```diff
@@ -22,3 +26,59 @@ def test_get_stats():
+    # sim and obs need to be Series
+    obs = obs[obs.columns[0]]
+    sim = sim[sim.columns[0]]
+    get_stats(sim, obs)
```
where is the assertion?

pmav99 commented on tests/compute_stats_test.py:
```diff
@@ -22,3 +26,59 @@ def test_get_stats():
+    # sim and obs need to be Series
+    obs = obs[obs.columns[0]]
+    sim = sim[sim.columns[0]]
+    storm_metrics(sim, obs, quantile=0.99, cluster=72)
```
Same to all of these. Where is the assertion?

pmav99 commented on seastats/stats.py:
```diff
@@ -214,28 +219,53 @@ def get_stats(
+        version_stat["madc"] = get_madc(sim, obs, round)
+    if should_calculate("kge"):
+        version_stat["kge"] = get_kge(sim, obs, round)
+
```
For each metric you are making two function calls instead of one (`should_calculate() + get_*()` VS just `get_*()`). 

In this specific case, performance should not be affected because the `get_*()` are probably dominating the runtime. That being said it is possible to refactor and and avoid the `should_calculate()`. 

A common way to avoid the consecutive `ifs` is to define a dictionary that maps the metrics name to the function calls. For example:

```python
_METRIC_TO_FUNC = {
    "mad": get_mad,
    "madp": get_madp,
    "madc": get_madc,
    "kge": get_kge,
}

version_stat = {}
for metric_name in metric_list:
    func = _METRIC_TO_FUNC[metric_name]
    version_stat[metric_name] = func(sim, obs, round)
```
The caveat here is that all the functions need to have the same signature. You can circumvent this by adding `*args, **kwargs` in all the functions, but it is a bit ugly IMHV.

Another way is to use the relatively new `match` statement which has beed added exactly for this use case. It should be something like this:
``` python
for metric_name in metrics_list:
    match metric_name:
        case "mad":
            version_stat[metric_name] = get_mad(sim, obs, round)
        case "madp":
            version_stat[metric_name] = get_madp(sim, obs, round)
        case "madc":
            version_stat[metric_name] = get_madc(sim, obs, round)
        case "kge":
            version_stat[metric_name] = get_kge(sim, obs, round)
````
In this specific case I would go for the match case, but it is only supported after 3.10

tomsail replied:
let's go for the `match` statement then. python 3.8 is end-of-life already and 3.9 ends in Oct 2025.

tomsail replied:
I updated CI for python versions above 3.10 included
It's indeed more elegant this way, thanks.

pmav99 commented on seastats/stats.py:
```diff
@@ -220,52 +243,67 @@ def get_stats(
-        return metrics is None or metric_name in metrics_list
+    if metrics == ["all"]:
+        metrics = _ALL_METRICS
+    print(metrics)
```
remove the `print()`

pmav99 commented on seastats/stats.py:
```diff
@@ -220,52 +243,67 @@ def get_stats(
+                else:
+                    slopepp, interceptpp = get_slope_intercept_pp(sim, obs, round)
+                    version_stat["slope_pp"] = slopepp
+                    version_stat["intercept_pp"] = interceptpp
```
I don't understand lines 278-299. Does `slope` make sense without `intercept`? If not, then I would just rename the metric to `slope_and_intercept`. If they make sense independently then I would do it like this:

``` python
            case "intercept":
                _, intercept = get_slope_intercept(sim, obs, round)
                version_stat["intercept"] = intercept
            case "slope":
                slope, _ = get_slope_intercept(sim, obs, round)
                version_stat["slope"] = slope
```
No reason to complicate the code just to avoid one function call which we are not even sure if it affects the runtimes

tomsail replied:
For the normal slope/intercept, it's okay indeed. 
But for the `*_pp` (percentiles) it does affect runtime. 
That's why I wanted to avoid to calculate it twice.

pmav99 commented on .github/workflows/run_tests.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
       matrix:
         os: ["ubuntu-latest"]
-        python: ["3.9", "3.10", "3.11", "3.12"]
+        python: ["3.10", "3.11", "3.12"]
```
Try to run 3.13, too. 3.13.0 has already been released. Some of our dependencies might be broken, but let's give it a try.

pmav99 commented on seastats/stats.py:
```diff
@@ -1,12 +1,35 @@
 import numpy as np
 import pandas as pd
 
+_ALL_METRICS = [
```
Let's rename this to `SUPPORTED_METRICS` and make it part of the public interface.

tomsail commented:
I have addressed the various things we've discussed yesterday. 
* made visible the `quantile` and `round` parameters. Both have default values (0 and -1) and deactivated.
* cleaned the tests and used `pytest.approx`
* using now by default `GENERAL_METRICS` and `STORM_METRIC` parameters lists for `get_stats()` and `storm_metrics()`. 
* vectorised the percentile calculation so it's now faster and made the `slope`/`intercept` API simpler

tomsail commented on seastats/__init__.py:
```diff
@@ -0,0 +1,7 @@
+from seastats.storms import STORM_METRICS_ALL
+
+SUGGESTED_METRICS = sorted(GENERAL_METRICS + STORM_METRICS)
+SUPPORTED_METRICS = sorted(GENERAL_METRICS_ALL + STORM_METRICS_ALL)
```
there's still a grey zone here. These lists are not used yet. 
The real added value would be to have a common and higher level function that uses `get_stats` and `storm_metrics` depending on the parameters that are queried.

pmav99 commented on seastats/__init__.py:
```diff
@@ -1,7 +1,196 @@
+        obs_ = obs[obs > obs.quantile(quantile)]
+    else:
+        sim_ = sim
+        obs_ = obs
```
Why are we renaming `obs` and `sim`? I probably miss something, but what is wrong with:
``` python
if quantile:
    sim = sim[sim > sim.quantile(quantile)]
    obs = obs[obs > obs.quantile(quantile)]
    
stats = {}
for metric in metrics:
    # ...
```

tomsail replied:
This is important, and I should have specified it. 
When calculating the storm events, with the PoT method, I don't think subsetting the TS before with the quantiles is a good idea. 
This would produce a fragmented time series that could interfere with the event detection.

tomsail replied:
That being said, putting the code 
```python
    # Storm metrics part with PoT Selection
    if np.any([m in STORM_METRICS_ALL for m in metrics]):
        df = match_extremes(sim, obs, quantile=quantile, cluster=cluster)
```
before the quantile selection, fixes the issue and simplefies also your review [here](https://github.com/oceanmodeling/seastats/pull/1#discussion_r1836653370) below

pmav99 commented on seastats/__init__.py:
```diff
@@ -1,7 +1,196 @@
+def get_stats(
+    sim: pd.Series,
+    obs: pd.Series,
+    metrics: Sequence[str] = None,
```
This type hint is obviously wrong. `None` is not a sequence of strings. Which brings up two points:
1. If None is equivalent to SUGGESTED_METRICS why not put `SUGGESTED_METRICS` as the default value?
2. We need to be running `mypy --strict` on CI.

tomsail replied:
1. Good point.
2. Adding mypy in the pre commit CI

pmav99 commented on seastats/__init__.py:
```diff
@@ -1,7 +1,196 @@
+
+    # Storm metrics part with PoT Selection
+    if np.any([m in STORM_METRICS_ALL for m in metrics]):
+        df = match_extremes(sim, obs, quantile=quantile, cluster=cluster)
```
maybe rename `df` to something more specific? E.g. `extremes_df` or something.

pmav99 commented on seastats/__init__.py:
```diff
@@ -1,7 +1,196 @@
+                case "error_norm":
+                    stats["error_norm"] = df["error_norm"].mean()
+    else:
+        logger.info("no storm metric specified")
```
Haven't tested it, but I think It should be possible to merge the two match:cases E.g. with something like this:
``` python
    if np.any([m in STORM_METRICS_ALL for m in metrics]):
        extremes_df = match_extremes(sim, obs, quantile=quantile, cluster=cluster)
    for metric in metrics:
        match metric:
            # General metrics
            case "bias":
                stats["bias"] = get_bias(sim_, obs_)
            # ...
            # Storm metrics
            case "R1":
                stats["R1"] = extremes_df["error"].iloc[0]
            # ...
````

Keep in mind that linters might complain that `extremes_df` may be undefined etc.

Not sure if it is better. Just writing it here so that you can consider it. Your call.

tomsail committed:
fix: adapt notebook and fix in list
